cleanup-passes/typescript.ts

```diff
@@ -0,0 +1,109 @@
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+'use strict';
+
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import {execFile} from 'child_process';
+import {promisify} from 'util';
+
+import {register} from '../cleanup-pass';
+import {ElementRepo} from '../element-repo';
+import {makeCommit} from './util';
+
+const generatorPackageName = '@polymer/gen-typescript-declarations';
+const generatorSemver = '^0.3.6';
+const npmScriptName = 'update-types';
+const npmScriptCommand =
+    'rm -f *.d.ts **/*.d.ts && gen-typescript-declarations --outDir .';
+
+const execFilePromise = promisify(execFile);
+
+/**
+ * This pass updates an element repo to include typings generated by
+ * https://github.com/Polymer/gen-typescript-declarations/ and configures an
+ * "update-types" NPM script that can be run to re-generate them.
+ *
+ * Throws an error if the repo has a missing or invalid package.json or if npm
+ * or bower aren't globally installed.
+ */
+async function typescriptPass(element: ElementRepo): Promise<void> {
+  const packageJsonPath = path.join(element.dir, 'package.json');
+  let packageJson: {
+    scripts?: {[key: string]: string},
+    devDependencies?: {[key: string]: string},
+  };
+  try {
+    packageJson = await fse.readJson(packageJsonPath);
+  } catch {
+    throw new Error(
+        `Missing or invalid package.json for element: ${element.ghRepo.name}`);
+  }
+
+  if (packageJson.devDependencies === undefined) {
+    packageJson.devDependencies = {};
+  }
+  if (packageJson.scripts === undefined) {
+    packageJson.scripts = {};
+  }
+
+  let writePackageJson = false;
+  if (packageJson.devDependencies[generatorPackageName] !== generatorSemver) {
+    packageJson.devDependencies[generatorPackageName] = generatorSemver;
+    writePackageJson = true;
+  }
+  if (packageJson.scripts[npmScriptName] !== npmScriptCommand) {
+    packageJson.scripts[npmScriptName] = npmScriptCommand;
+    writePackageJson = true;
+  }
+  if (writePackageJson) {
+    await fse.writeJson(packageJsonPath, packageJson, {spaces: 2});
+  }
+
+  const execOpts = {cwd: element.dir};
+  await execFilePromise('npm', ['install'], execOpts);
+  // We also have to bower install because the generator needs to see the
+  // dependencies.
+  await execFilePromise('bower', ['install'], execOpts);
+  await execFilePromise('npm', ['run', npmScriptName], execOpts);
+
+  let doCommit = false;
+  const commitFiles = [];
+  for (const changedFile of await element.repo.getStatus()) {
+    const filepath = changedFile.path();
+    if (filepath.endsWith('.d.ts')) {
+      doCommit = true;
+      commitFiles.push(filepath);
+    } else if (filepath === 'package.json') {
+      doCommit = true;
+      commitFiles.push(filepath);
+    } else if (filepath === 'package-lock.json') {
+      // If all we did was update the package lock, don't bother with the
+      // commit. But do include it if we changed something else.
+      commitFiles.push(filepath);
+    }
```

Comment:
do we expect other file paths? If not, should we log unexpected ones?

Reply:
Added a console.log to point these out.